consensus/core/src/transaction_certifier.rs

```diff
@@ -0,0 +1,814 @@
+    ) -> Vec<CertifiedBlock> {
+        let mut certified_blocks = vec![];
+        for (voted_block, reject_txn_votes) in voted_blocks {
+            let blocks = self.add_voted_block(voted_block, reject_txn_votes);
+            certified_blocks.extend(blocks);
+        }
+        certified_blocks
+    }
+
+    fn add_voted_block(
+        &mut self,
+        voted_block: VerifiedBlock,
+        reject_txn_votes: Vec<TransactionIndex>,
+    ) -> Vec<CertifiedBlock> {
+        let mut certified_blocks = vec![];
+
+        if voted_block.round() <= self.gc_round {
+            // Block is outside of GC bound.
+            return vec![];
+        }
+
+        let vote_info = self.votes.entry(voted_block.reference()).or_default();
+        if vote_info.block.is_some() {
+            // Input block has already been processed and added to the state.
+            return vec![];
+        }
+        vote_info.block = Some(voted_block.clone());
+        vote_info.own_reject_txn_votes = Some(reject_txn_votes.clone());
+
+        // Add own reject txn votes for the block.
+        for reject in &reject_txn_votes {
+            vote_info
+                .reject_txn_votes
+                .entry(*reject)
+                .or_default()
+                .add_unique(self.context.own_index, &self.context.committee);
+        }
+
+        // Add own accept votes for the block.
+        vote_info
+            .accept_block_votes
+            .add_unique(self.context.own_index, &self.context.committee);
+
+        // Check if the input block is now certified after the updates above.
+        // NOTE: votes can already exist for the block and its transactions.
+        if let Some(certified_block) = vote_info.take_certified_output(&self.context.committee) {
+            certified_blocks.push(certified_block);
+        }
+
+        // Update reject votes from the block first. Otherwise another block can get certified when it should not be.
+        for block_votes in voted_block.transaction_votes() {
+            if block_votes.block_ref.round <= self.gc_round {
+                // Block is outside of GC bound.
+                continue;
+            }
+            let vote_info = self.votes.entry(block_votes.block_ref).or_default();
+            for reject in &block_votes.rejects {
+                vote_info
+                    .reject_txn_votes
+                    .entry(*reject)
+                    .or_default()
+                    .add_unique(voted_block.author(), &self.context.committee);
+            }
+            if let Some(certified_block) = vote_info.take_certified_output(&self.context.committee)
+            {
+                certified_blocks.push(certified_block);
+            }
+        }
+
+        // Update accept votes from the block after updating reject votes.
+        // Only parent round blocks receive accept votes on the fastpath.
+        for ancestor in voted_block.ancestors() {
```

Comment:
Is there a scenario where ancestors and the Vec<BlockTransactionVotes> returned from transaction_votes() don't line up? Should we loop through transaction_votes()?